### PR TITLE
feat: Add "long" formatted CSV output.

### PR DESCRIPTION
**Pull request by jetuk**
Add an output to pywr-core for writing long formatted CSV files. This is supported through the existing CSV output type in the schema.

### CODE REVIEW

File: pywr-core/src/recorders/csv.rs [hunk 1]
```diff
@@ -21,17 +21,57 @@ struct Internal {
...
         }
     }
+
+    fn write_values(
+        &self,
+
+        metric_set_states: &[Vec<MetricSetState>],
+        internal: &mut Internal,
+    ) -> Result<(), PywrError> {
+        let mut row = Vec::new();
```
Comment (Batch21) [inline pywr-core/src/recorders/csv.rs:39]: Can this be given a capacity ?

File: pywr-schema/src/outputs/csv.rs [hunk 2]
```diff
@@ -65,9 +117,11 @@ mod tests {
...
         model.run::<ClpSolver>(&ClpSolverSettings::default()).unwrap();
 
-        // After model run there should be an output file.
-        let expected_path = temp_dir.path().join("outputs.csv");
-        assert!(expected_path.exists());
+        // After model run there should be two output files.
+        let expected_long_path = temp_dir.path().join("outputs-long.csv");
+        assert!(expected_long_path.exists());
+        let expected_wide_path = temp_dir.path().join("outputs-wide.csv");
+        assert!(expected_wide_path.exists());
```
Comment (Batch21) [inline pywr-schema/src/outputs/csv.rs:124]: Would it be worth loading the outputs csv in these tests and doing some basic checks (e.g. shape, headers, etc..)?

**Review by Batch21 — COMMENTED** [review]
These work well and its nice to have the choice of format. Made a few suggestions

File: pywr-schema/src/outputs/csv.rs [hunk 3]
```diff
@@ -20,10 +41,37 @@ impl CsvOutput {
             _ => self.filename.to_path_buf(),
         };
 
-        let metric_set_idx = network.get_metric_set_index_by_name(&self.metric_set)?;
-        let recorder = CSVRecorder::new(&self.name, filename, metric_set_idx);
+        let recorder: Box<dyn Recorder> = match self.format {
+            CsvFormat::Wide => match &self.metric_set {
```
[the remark this answers is not in the document]
Comment (jetuk) [reply pywr-schema/src/outputs/csv.rs:51]: Done with an untagged enum.

File: pywr-schema/src/outputs/csv.rs (hunk 2 above)
Comment (jetuk) [reply pywr-schema/src/outputs/csv.rs:148]: Done.

**Review by Batch21 — APPROVED** [review]

**Pull request merged**